tools/walletextension/subscriptions/deduplication_circular_buffer.go

```diff
@@ -0,0 +1,43 @@
+package subscriptions
+
+import "github.com/ethereum/go-ethereum/common"
+
+// LogKey uniquely represents a log (consists of BlockHash, TxHash, and Index)
+type LogKey struct {
+	BlockHash common.Hash // Not necessary, but can be helpful in edge case of block reorg.
+	TxHash    common.Hash
+	Index     uint
+}
+
+// CircularBuffer is a data structure that uses a single, fixed-size buffer as if it was connected end-to-end.
+type CircularBuffer struct {
+	data []LogKey
+	size int
+	end  int
+}
+
+// NewCircularBuffer initializes a new CircularBuffer of the given size.
+func NewCircularBuffer(size int) *CircularBuffer {
```

Comment:
Why a circular buffer versus a limitless slice ?

Reply:
What happens to the user experience if the buffer starts removing messages ?

Perhaps it would be simpler from a user experience and system operation if the subscription is knocked down when the buffer reaches a certain size ?

Reply:
Sorted offline 👍